File: push/src/main/java/top/trumeet/mipush/provider/event/type/TypeFactory.java
package top.trumeet.mipush.provider.event.type;

import android.annotation.SuppressLint;

import com.nihility.XMPushUtils;
import com.xiaomi.xmpush.thrift.ActionType;
import com.xiaomi.xmpush.thrift.XmPushActionContainer;
import com.xiaomi.xmsf.utils.ConvertUtils;

import top.trumeet.mipush.provider.entities.Event;
import top.trumeet.mipush.provider.event.EventType;

/**
 * Created by Trumeet on 2018/2/7.
 */

public class TypeFactory {
    public static EventType createForStore(XmPushActionContainer buildContainer) {
        String pkg = buildContainer.packageName;
        String info = String.valueOf(ConvertUtils.toJson(buildContainer));
        byte[] payload = XMPushUtils.packToBytes(buildContainer);

        switch (buildContainer.getAction()) {
            case SendMessage:
                NotificationType eventType = new NotificationType(info, pkg, payload);
                eventType.setType(Event.Type.SendMessage);
                return eventType;
            case Notification:
                return new NotificationType(info, pkg, payload);
            case Registration:
                return new RegistrationResultType(info, pkg, payload);
            default:
                return new UnknownType(getTypeId(buildContainer.getAction()), info, pkg, payload);
        }
    }

    public static EventType createForDisplay(Event eventFromDB) {
        String pkg = eventFromDB.getPkg();
        String info = eventFromDB.getInfo();
        byte[] payload = eventFromDB.getPayload();

        switch (eventFromDB.getType()) {
            case Event.Type.Command:
                return new CommandType(info, pkg, payload);
            case Event.Type.Notification:
                return new NotificationType(info, pkg, payload);
            case Event.Type.SendMessage:
                NotificationType type = new NotificationType(info, pkg, payload);
                type.setType(Event.Type.SendMessage);
                return type;
            case Event.Type.Registration:
                return new RegistrationType(info, pkg, payload);
            case Event.Type.RegistrationResult:
                return new RegistrationResultType(info, pkg, payload);
            default:
                return new UnknownType(eventFromDB.getType(), info, pkg, payload);
        }
    }


    @SuppressLint("WrongConstant")
    private static @Event.Type int getTypeId(ActionType type) {
        switch (type) {
            case Command:
                return Event.Type.Command;
            case SendMessage:
                return Event.Type.SendMessage;
            case Notification:
                return Event.Type.Notification;
            case SetConfig:
                return Event.Type.SetConfig;
            case AckMessage:
                return Event.Type.AckMessage;
            case Registration:
                return Event.Type.Registration;
            case Subscription:
                return Event.Type.Subscription;
            case ReportFeedback:
                return Event.Type.ReportFeedback;
            case UnRegistration:
                return Event.Type.UnRegistration;
            case UnSubscription:
                return Event.Type.UnSubscription;
            case MultiConnectionResult:
                return Event.Type.MultiConnectionResult;
            case MultiConnectionBroadcast:
                return Event.Type.MultiConnectionBroadcast;
            default:
                return -1;
        }
    }
}

File: push/src/main/java/io/github/magisk317/mipush/runtime/store/event/type/TypeFactory.kt
package io.github.magisk317.mipush.runtime.store.event.type

import android.annotation.SuppressLint
import com.magisk317.XMPushUtils
import com.xiaomi.xmpush.thrift.ActionType
import com.xiaomi.xmpush.thrift.XmPushActionContainer
import com.xiaomi.xmsf.utils.ConvertUtils
import io.github.magisk317.mipush.runtime.store.entities.Event
import io.github.magisk317.mipush.runtime.store.event.EventType

/**
 * Created by Trumeet on 2018/2/7.
 */
object TypeFactory {
    @JvmStatic
    fun createForStore(buildContainer: XmPushActionContainer): EventType {
        val pkg = buildContainer.packageName
        val info = ConvertUtils.toJson(buildContainer).toString()
        val payload = XMPushUtils.packToBytes(buildContainer)

        return when (buildContainer.getAction()) {
            ActionType.SendMessage -> {
                val eventType = NotificationType(info, pkg, payload)
                eventType.type = Event.Type.SendMessage
                eventType
            }
            ActionType.Notification -> NotificationType(info, pkg, payload)
            ActionType.Registration -> RegistrationResultType(info, pkg, payload)
            else -> UnknownType(getTypeId(buildContainer.getAction()), info, pkg, payload)
        }
    }

    @JvmStatic
    fun createForDisplay(eventFromDB: Event): EventType {
        val pkg = eventFromDB.pkg
        val info = eventFromDB.info
        val payload = eventFromDB.payload

        return when (eventFromDB.type) {
            Event.Type.Command -> CommandType(info, pkg, payload)
            Event.Type.Notification -> NotificationType(info, pkg, payload)
            Event.Type.SendMessage -> {
                val type = NotificationType(info, pkg, payload)
                type.type = Event.Type.SendMessage
                type
            }
            Event.Type.Registration -> RegistrationType(info, pkg, payload)
            Event.Type.RegistrationResult -> RegistrationResultType(info, pkg, payload)
            else -> UnknownType(eventFromDB.type, info, pkg, payload)
        }
    }

    @SuppressLint("WrongConstant")
    @JvmStatic
    @Event.Type
    private fun getTypeId(type: ActionType): Int {
        return when (type) {
            ActionType.Command -> Event.Type.Command
            ActionType.SendMessage -> Event.Type.SendMessage
            ActionType.Notification -> Event.Type.Notification
            ActionType.SetConfig -> Event.Type.SetConfig
            ActionType.AckMessage -> Event.Type.AckMessage
            ActionType.Registration -> Event.Type.Registration
            ActionType.Subscription -> Event.Type.Subscription
            ActionType.ReportFeedback -> Event.Type.ReportFeedback
            ActionType.UnRegistration -> Event.Type.UnRegistration
            ActionType.UnSubscription -> Event.Type.UnSubscription
            ActionType.MultiConnectionResult -> Event.Type.MultiConnectionResult
            ActionType.MultiConnectionBroadcast -> Event.Type.MultiConnectionBroadcast
            else -> -1
        }
    }
}
